### DISCUSSION TITLE
Best Practices for Updating Shared Map on Control Flow Instructions

### DISCUSSION THREAD

**Discussion by kevin-valerio** [wasmi-labs/wasmi#1143] [Q&A] [answered]
Hello,

I am making some experimentations with WASMI. My goal is to update a shared map whenever a new control flow instruction is hit, so that I could use that map in my fuzzer to fetch the coverage.

I am uncertain about the optimal location to update this map. One potential spot I identified is [push_frame](https://github.com/wasmi-labs/wasmi/blob/7b6efbbfa07c88d775881f8d87ffdd98494105ac/crates/wasmi/src/engine/translator/control_stack.rs#L68), but I would appreciate any opinions or recommendations on this approach.

Thanks! :)

**Comment by Robbepop** [comment]
Hi @kevin-valerio , what exactly do you mean with "shared map"? Translation? Execution? Anything else?
And what exactly do you mean with "a new control flow instruction is hit"?

**Reply by kevin-valerio** [reply]
Hey @Robbepop ! 

Thank you very much for your answer.

My objective is to obtain coverage data of a WASM blob during its execution, for fuzzing purposes. Once obtained, this data will be shared via a shared map to the fuzzer, but the sharing part is not the main concern right now.

To achieve this, I need to patch WASMI so that on each new control flow instruction, I can update my coverage map. By control flow instruction, I mean every WASM instruction that leads to a new edge in the control flow graph, for instance `br`, `br_if`, `br_table`, `if`, `else`, `call`, `return`, `loop`, `block`, `end`.

And my question is, where would be the most appropriate part of the code to do so.
Where in WASMI should I introduce my logic in order to update this coverage map (a simple `[branch_id <-> hit_count]`) ?

I hope my question makes more sense, let me know if you need more clarifications

**Reply by Robbepop** [reply]
So am I right that you want to track when an `if`, `block` and `loop` control flow is entered and existed during execution for coverage tracking purposes?
If that is the case then you are in for some more trouble than you probably thought because all this information is lost during the Wasm bytecode to Wasmi bytecode translation phase.
Wasmi does not execute Wasm bytecode, it executes Wasmi bytecode. This Wasmi bytecode is the result of the translation from Wasm bytecode to Wasmi bytecode in the Wasmi translator which is run before a Wasm function is executed.
There is no more notion of `block`, `loop`, `if` or `end` in Wasmi bytecode, thus it is not easily possible to track those during execution. If you want to preserve this information you probably need to adjust Wasmi bytecode (and Wasmi translator) to not lose this information.
This could be done by re-adding `if`, `block`, `loop` and `end` instructions (with their respective meta data) to the Wasmi bytecode and preserve them during translation. They should be implemented as no-ops besides the stuff you want to do.

**Reply by kevin-valerio** [reply]
I see. I wasn't aware of this.

But even if those instructions are lost, it doesn't matter, I would just need to detect when WASMI engine is entering a new branch, right, and then update my map from there. The point is only to track the WASM coverage, even if those specific instructions aren't there after translation.

**Reply by Robbepop** [reply]
As I have previously mentioned, not only the instructions are lost but all information associated to Wasm control flow is lost. So you need to preserve it if you want to use this information during Wasm execution. Wasmi bytecode only has low-level branching instructions such as `br`, `br_le` etc.

**Reply by kevin-valerio** [reply]
As far as I understand, any branch redirection will ultimately call [`branch_to`](https://github.com/kevin-valerio/wasmi/blob/96aa797c28a670798bd4cd87e654636b794561a9/crates/wasmi/src/engine/executor/instrs/branch.rs#L26), why couldn't that be used a hook to track branch flow?

**Reply by Robbepop** [reply, marked as answer]
It could technically be used but you wouldnt know if you came or exited from a Wasm if, block or loop control flow construct. If that suffices your needs you should be fine.

**Reply by kevin-valerio** [reply]
Thanks for your help :)
